publishing-bot/Makefile

```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+all: build
+.PHONY: all
+
+-include $(CONFIG)
+-include $(CONFIG)-token
+
+DOCKER_REPO ?= k8s-publishing-bot
+NAMESPACE ?=
+KUBECTL ?= kubectl
+SCHEDULE ?= 0 5 * * *
+INTERVAL ?= 86400
+MEMORY_REQUESTS ?= 200Mi
+MEMORY_LIMITS ?= 1.6Gi
+GOOS ?= linux
+
+build_cmd = GO111MODULE=on mkdir -p _output && GOOS=$(GOOS) go build -o _output/$(1) ./cmd/$(1)
+prepare_spec = sed 's,DOCKER_IMAGE,$(DOCKER_REPO),g;s,MEMORY_REQUESTS,$(MEMORY_REQUESTS),g;s,MEMORY_LIMITS,$(MEMORY_LIMITS),g'
+
+SHELL := /bin/bash
+
+build:
+	$(call build_cmd,collapsed-kube-commit-mapper)
+	$(call build_cmd,publishing-bot)
+	$(call build_cmd,sync-tags)
+	$(call build_cmd,init-repo)
+	$(call build_cmd,godeps-gen)
+	$(call build_cmd,gomod-zip)
+.PHONY: build
+
+build-image: build
+	docker build -t $(DOCKER_REPO) .
+.PHONY: build-image
+
+push-image:
+	docker push $(DOCKER_REPO):latest
+
+clean:
+	rm -rf _output
+.PHONY: clean
+
+update-deps:
+	glide update --strip-vendor
+.PHONY: update-deps
+
+validate:
+	if [ -f $(CONFIG)-rules-configmap.yaml ]; then \
+		go run ./cmd/validate-rules <(sed '1,/config: /d;s/^    //' $(CONFIG)-rules-configmap.yaml); \
+	else \
+		go run ./cmd/validate-rules $$(grep "rules-file: " $(CONFIG)-configmap.yaml | sed 's/.*rules-file: //'); \
+	fi
+.PHONY: validate
+
+init-deploy: validate
+	$(KUBECTL) delete -n "$(NAMESPACE)" --ignore-not-found=true replicaset publisher
+	$(KUBECTL) delete -n "$(NAMESPACE)" --ignore-not-found=true pod publisher
+	while $(KUBECTL) get pod -n "$(NAMESPACE)" publisher -a &>/dev/null; do echo -n .; sleep 1; done
+	$(KUBECTL) apply -n "$(NAMESPACE)" -f manifests/storage-class.yaml || true
+	$(KUBECTL) get StorageClass ssd
+	$(KUBECTL) apply -n "$(NAMESPACE)" -f $(CONFIG)-configmap.yaml
+	$(KUBECTL) apply -n "$(NAMESPACE)" -f $(CONFIG)-rules-configmap.yaml; \
+	$(KUBECTL) apply -n "$(NAMESPACE)" -f manifests/pvc.yaml
+
+run: init-deploy
+	{ cat manifests/pod.yaml && sed 's/^/  /' manifests/podspec.yaml; } | \
+	$(call prepare_spec) | $(KUBECTL) apply -n "$(NAMESPACE)" -f -
+
+deploy: init-deploy
+	$(KUBECTL) apply -n "$(NAMESPACE)" -f manifests/service.yaml
+	{ cat manifests/rs.yaml && sed 's/^/      /' manifests/podspec.yaml; } | \
+	$(call prepare_spec) | sed 's/-interval=0/-interval=$(INTERVAL)/g' | \
+	$(KUBECTL) apply -n "$(NAMESPACE)" -f -
```

Comment:
why are we duplicating deployment code from the publishing bot? 😱 This repo is invisible to anybody doing changes in the bot repository, which will eventually break this copy.

Reply:
With the migration of the publishing bot to a new community-owned GKE cluster (`aaa`), we want to migrate some parts of the deployment code. What's the right approach in this scenario ?

Reply:
This code in the bot repo is used by a number of parties. So we won't remove it. Hence, you are creating a copy here which will most certainly divert. 

Where is the problem of keeping the deployment code where it is? The people deploying the publishing bot (which is @nikhita, @dims and myself only; nobody without deep knowledge of what this does should do this) all are used to the bot repo. Which pain points should this duplication solve?

I would also be fine to do some git clone magic here checking out the bot repo and call the Makefile in there. Then it's well documented what is going on.

Reply:
I am fine to leave the config in the bot repo, I guess.  This repo/dir could be reduced to a README explaining how we deploy it in "aaa"

Reply:
But all my comments here apply - especially around CPu and memory limits.

Reply:
@ameukam can you move the fixes to the publishing-bot repo instead? @nikhita and I are happy to review over there. Thanks!

Reply:
@sttts @thockin @ameukam @nikhita I have added to todays agenda for k8s-infra-wg bi-weekly meeting discussion about this topic. Please attend if you want to discuss about it in syncronous manner. http://bit.ly/wg-k8s-infra-notes